core/ecschema-metadata/src/SchemaFormatsProvider.ts

```diff
@@ -0,0 +1,144 @@
+   */
+  constructor(contextOrLocater: ISchemaLocater) {
+    if (contextOrLocater instanceof SchemaContext) {
+      this._context = contextOrLocater;
+    } else {
+      this._context = new SchemaContext();
+      this._context.addLocater(contextOrLocater);
+    }
+  }
+
+  /**
+   * Try retrieving the format from schemas. If the format is not part of the base Formats schema, but from a KindOfQuantity, the format will be constructed
+   * from the default presentation format of that KindOfQuantity.
+   */
+  private async getFormatFromSchema(id: string, unitSystem?: UnitSystemKey): Promise<SchemaItemFormatProps | undefined> {
+    const [schemaName, schemaItemName] = SchemaItem.parseFullName(id);
+    const schemaKey = new SchemaKey(schemaName);
+    const schema = await this._context.getSchema(schemaKey);
+    if (!schema) {
+      return undefined;
+    }
+    const itemKey = new SchemaItemKey(schemaItemName, schema.schemaKey);
+
+    if (schema.name === "Formats") {
+      const format = await this._context.getSchemaItem(itemKey, Format);
+      if (!format) {
+        return undefined;
+      }
+      return format.toJSON(true);
+    }
+    return this.getKindOfQuantityFormatFromSchema(itemKey, unitSystem);
+  }
+
+  private async getKindOfQuantityFormatFromSchema(itemKey: SchemaItemKey, unitSystem?: UnitSystemKey): Promise<SchemaItemFormatProps | undefined> {
+    const kindOfQuantity = await this._context.getSchemaItem(itemKey, KindOfQuantity);
+
+    if (!kindOfQuantity) {
+      return undefined;
+    }
+    // Find the first presentation format that matches the provided unit system. Else, use the first entry.
+    if (unitSystem) {
+      const unitSystemGroupNames = getUnitSystemGroupNames(unitSystem);
+      const presentationFormats = kindOfQuantity.presentationFormats;
+      for (const system of unitSystemGroupNames) {
+        for (const format of presentationFormats) {
+          const unit = format.units && format.units[0][0];
+          if (!unit) {
+            continue;
+          }
+          const currentUnitSystem = await unit.unitSystem;
+          if (currentUnitSystem && currentUnitSystem.name.toUpperCase() === system) {
+            return getFormatProps(format);
+          }
+        }
+      }
+      return undefined;
+    }
+    const defaultFormat = kindOfQuantity.defaultPresentationFormat;
+    if (!defaultFormat) {
+      return undefined;
+    }
+    return getFormatProps(defaultFormat);
+  }
+
+  /**
+   * Retrieves a Format from the cache or from the schema. If retrieving from the schema, and the format is part of a KindOfQuantity,
+   * an optional UnitSystemKey can be provided to get the first presentation format in the KindOfQuantity that matches the unit system.
+   * @param id The full name of the Format.
+   * @returns
+   */
+  public async getFormat(id: string, unitSystem?: UnitSystemKey): Promise<SchemaItemFormatProps | undefined> {
+    if (this._formatCache.has(id)) {
```

Comment:
I think it would be great to normalize the given id for caching purposes. You could be given all kinds of different variations of the same thing: `MySchema.MyKoQ`, `MySchema:MyKoQ`, `myschema.mykoq` and so on.